File: src/test/java/com/bemsa/scraper/clients/GithubClientTest.java
package com.bemsa.scraper.clients;

import com.bemsa.scraper.clients.impl.GithubClientImpl;
import com.bemsa.scraper.exceptions.GitScraperException;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestClientException;
import org.springframework.web.client.RestTemplate;

import static com.bemsa.scraper.constants.TestConstants.*;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class GithubClientTest {

    @Mock
    private RestTemplate restTemplate;

    @InjectMocks
    private GithubClientImpl githubClient;

    @DisplayName("GET request returns 200")
    @Test
    void testGet() {
        //given
        String json = "{}";
        ResponseEntity<String> userResponse = ResponseEntity.ok(json);
        when(getResponseEntity()).thenReturn(userResponse);

        //when
        ResponseEntity<String> request = githubClient.get(API_URL, USER_NAME);

        //then
        assertThat(request)
                .isNotNull()
                .isEqualTo(userResponse);
    }

    @DisplayName("GET request returns 4xx")
    @Test
    void testGetThrowsRestClientException() {
        assertThatThrownBy(() -> {
            //given
            when(getResponseEntity()).thenThrow(RestClientException.class);

            //when
            var response = githubClient.get(API_URL, USER_NAME);

            //then
            assertThat(response.getStatusCode())
                    .isEqualTo(HttpStatus.BAD_REQUEST);
        }).isInstanceOf(GitScraperException.class);
    }

    @DisplayName("GET request returns 5xx")
    @Test
    void testGetThrowsException() {
        assertThatThrownBy(() -> {
            //given
            when(getResponseEntity()).thenThrow(NullPointerException.class);

            //when
            var response = githubClient.get(API_URL, USER_NAME);

            //then
            assertThat(response).isNull();
        }).isInstanceOf(NullPointerException.class);
    }

    @Test
    void testAsJson() {
        //given
        String json = "{}";
        ResponseEntity<String> userResponse = ResponseEntity.ok(json);
        when(getResponseEntity()).thenReturn(userResponse);

        //when
        String requestJson = githubClient.asJson(USER_NAME, API_URL, ERROR_MESSAGE);

        //then
        assertThat(requestJson)
                .isEqualTo(json);
    }

    @Test
    void testGetRequestAsJsonThrowsException() {
        assertThatThrownBy(() -> {
            //given
            when(getResponseEntity()).thenThrow(RestClientException.class);

            //when
            String requestJson = githubClient.asJson(USER_NAME, API_URL, ERROR_MESSAGE);

            //then
            assertThat(requestJson).isNull();
        }).isInstanceOf(GitScraperException.class);
    }

    private ResponseEntity<String> getResponseEntity() {
        return restTemplate.exchange(
                anyString(),
                eq(HttpMethod.GET),
                any(HttpEntity.class),
                eq(String.class),
                anyString());
    }
}

File: src/test/kotlin/com/bemsa/scraper/clients/GithubClientTest.kt
package com.bemsa.scraper.clients;

import com.bemsa.scraper.clients.impl.GithubClientImpl
import com.bemsa.scraper.constants.API_URL
import com.bemsa.scraper.constants.USER_NAME
import com.ninjasquad.springmockk.MockkBean
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.boot.test.web.server.LocalServerPort
import org.springframework.http.*
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.web.client.RestTemplate

//import static com.bemsa.scraper.constants.TestConstants.*;
//import static org.assertj.core.api.Assertions.assertThat;
//import static org.assertj.core.api.Assertions.assertThatThrownBy;
//import static org.mockito.ArgumentMatchers.*;
//import static org.mockito.Mockito.when;

@ExtendWith(SpringExtension::class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class GithubClientTest {
    @MockkBean
    private lateinit var restTemplate: RestTemplate

//    @Autowired
//    private lateinit var restTemplate: TestRestTemplate

    @InjectMockKs
    private lateinit var githubClient: GithubClientImpl

//    private val githubClient = GithubClientImpl(restTemplate)

    @BeforeEach
    fun setUp() = MockKAnnotations.init(this)

    @DisplayName("GET request returns 200")
    @Test
    fun testGet() {
        //given
        val json = "{}"
        val userResponse = ResponseEntity.ok(json);
        every { getResponseEntity() } returns userResponse
//        val json = "{}"
//        ResponseEntity<String> userResponse = ResponseEntity.ok(json);
//        when(getResponseEntity(USER_NAME)).thenReturn(userResponse);

        //when
        val request = githubClient.get(API_URL, USER_NAME);

        //then
          verify(exactly = 1) { getResponseEntity()  }
          assertThat(request)
                .isNotNull
                .isEqualTo(userResponse);
    }

    @DisplayName("GET request returns 4xx")
    @Test
    fun testGetThrowsRestClientException() {
//        assertThatThrownBy(() -> {
//            //given
//            when(getResponseEntity()).thenThrow(RestClientException.class);
//
//            //when
//            var response = githubClient.get(API_URL, USER_NAME);
//
//            //then
//            assertThat(response.getStatusCode())
//                    .isEqualTo(HttpStatus.BAD_REQUEST);
//        }).isInstanceOf(GitScraperException.class);
    }

    @DisplayName("GET request returns 5xx")
    @Test
    fun testGetThrowsException() {
//        assertThatThrownBy(() -> {
//            //given
//            when(getResponseEntity()).thenThrow(NullPointerException.class);
//
//            //when
//            var response = githubClient.get(API_URL, USER_NAME);
//
//            //then
//            assertThat(response).isNull();
//        }).isInstanceOf(NullPointerException.class);
    }

    @Test
    fun testAsJson() {
//        //given
//        String json = "{}";
//        ResponseEntity<String> userResponse = ResponseEntity.ok(json);
//        when(getResponseEntity()).thenReturn(userResponse);
//
//        //when
//        String requestJson = githubClient.asJson(USER_NAME, API_URL, ERROR_MESSAGE);
//
//        //then
//        assertThat(requestJson)
//                .isEqualTo(json);
    }

    @Test
    fun testGetRequestAsJsonThrowsException() {
//        assertThatThrownBy(() -> {
//            //given
//            when(getResponseEntity()).thenThrow(RestClientException.class);
//
//            //when
//            String requestJson = githubClient.asJson(USER_NAME, API_URL, ERROR_MESSAGE);
//
//            //then
//            assertThat(requestJson).isNull();
//        }).isInstanceOf(GitScraperException.class);
    }

    private fun getResponseEntity(): ResponseEntity<String> = restTemplate.exchange(API_URL, HttpMethod.GET, getHttpEntity(), String::class.java, USER_NAME)

    private fun getHttpEntity(): HttpEntity<String> {
        val headers = HttpHeaders()
        headers.set("Accept", "application/vnd.github+json")
        headers.set("Cache-Control", "public, max-age=60, s-maxage=60")
        return HttpEntity(headers)
    }


//"",
//HttpMethod.GET),
//any(HttpEntity::class),
//eq(String.class),
//anyString());
}
